File: minimum_scalar_product.go
package main

import (
	"sort"
	"fmt"
	"strings"
	"bufio"
	"log"
	"os"
	"strconv"
)

func GrabLines(args string) []string {
	var lines []string
	file, err := os.Open(args)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines
}

func ToIntSlice(va []string) []int {
	var intslice []int
	for _, s := range va {
		newint, _ := strconv.Atoi(s)
		intslice = append(intslice, newint) 
	}
	return intslice
}

func ScalarProduct(va []int, vb []int, i int, numints int) {
	total := 0
	sort.Sort(sort.Reverse(sort.IntSlice(va)))
	sort.Sort(sort.IntSlice(vb))
	for i, e := range va { 
		total += e * vb[i]
	}
	fmt.Printf("Case #%d: ", i)
	fmt.Println(total)
}

func main() {
	argsWithoutProgram := os.Args[1]
	lines := GrabLines(argsWithoutProgram)
	cases, _ := strconv.Atoi(lines[0])
	index := 1
	for i := 1; i <= cases; i ++ {
		numints, _ := strconv.Atoi(lines[index])
		va := strings.Split(lines[index+1], " ")
		vb := strings.Split(lines[index+2], " ")
		newva := ToIntSlice(va)
		newvb := ToIntSlice(vb)
		ScalarProduct(newva, newvb, i, numints)
		index += 3
	}
}

File: minimum_scalar_product.rb
class MinimumScalar

def initialize(va, vb)
  @va = va.sort.reverse
  @vb = vb.sort
end

def va
  @va
end

def vb
  @vb
end

def solve
  total = 0
  va.each_with_index do |a, i|
    total += a * vb[i]
  end
  total
end

end

txt = open(ARGV.first)
lines = txt.readlines
cases = lines[0].to_i
i = 1
index = 0
cases.times do
  index += 1
  va = lines[i+1].split.map {|i| i.to_i}
  vb = lines[i+2].split.map {|i| i.to_i}
  @product = MinimumScalar.new(va, vb)
  puts "Case #{index}: #{@product.solve}"
  i += 3
end


